flink-connectors/flink-connector-datagen/src/test/java/org/apache/flink/connector/datagen/source/DataGeneratorSourceTest.java

```diff
@@ -99,23 +99,28 @@ void testReaderCheckpoints() throws Exception {
                         new NumberSequenceSource.NumberSequenceSplit("split-1", from, mid),
                         new NumberSequenceSource.NumberSequenceSplit("split-2", mid + 1, to)));
 
-        long remainingInCycle = elementsPerCycle;
-        while (reader.pollNext(out) != InputStatus.END_OF_INPUT) {
-            if (--remainingInCycle <= 0) {
-                remainingInCycle = elementsPerCycle;
-                // checkpoint
-                List<NumberSequenceSource.NumberSequenceSplit> splits = reader.snapshotState(1L);
-
-                // re-create and restore
-                reader = createReader();
-                if (splits.isEmpty()) {
-                    reader.notifyNoMoreSplits();
-                } else {
-                    reader.addSplits(splits);
-                }
+        for (int cycle = 0; cycle < 3; cycle++) {
+            // this call is not required but mimics what happens at runtime
+            reader.pollNext(out);
+            for (int elementInCycle = 0; elementInCycle < elementsPerCycle; elementInCycle++) {
+                reader.isAvailable().get();
+                reader.pollNext(out);
```

Comment:
```suggestion
            assertThat(reader.pollNext(out))
                    .as(
                            "Each poll should return a NOTHING_AVAILABLE status to explicitly trigger the availability check through in SourceReader.isAvailable")
                    .isEqualTo(InputStatus.NOTHING_AVAILABLE);
            for (int elementInCycle = 0; elementInCycle < elementsPerCycle; elementInCycle++) {
                reader.isAvailable().get();
                assertThat(reader.pollNext(out))
                        .as(
                                "Each poll should return a NOTHING_AVAILABLE status to explicitly trigger the availability check through in SourceReader.isAvailable")
                        .isEqualTo(InputStatus.NOTHING_AVAILABLE);
```
I feel like we should also add assertion here with a description to reason why the `NOTHING_AVAILABLE` is expected. Otherwise, it's kind of unintuitive that `NOTHING_AVAILABLE` is returned even though there is actually data available. WDYT?